engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCache.scala

```diff
@@ -0,0 +1,60 @@
+package cromwell.engine.workflow.lifecycle.execution.callcaching
+
+import cromwell.backend.BackendJobExecutionActor.SucceededResponse
+import cromwell.core.{JobOutputs, WorkflowId}
+import cromwell.database.sql._
+import cromwell.database.sql.tables.CallCachingResultMetainfoEntry
+import cromwell.engine.workflow.lifecycle.execution.callcaching.EngineJobHashingActor.CallCacheHashes
+import cromwell.core.ExecutionIndex.IndexEnhancedIndex
+import wdl4s.values._
+import language.postfixOps
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.Try
+
+/**
+  * Given a database-layer CallCacheStore, this accesser can access the database with engine-friendly data types.
+  */
+class CallCache(database: CallCachingStore) {
+  def addToCache(workflowId: WorkflowId, callCacheHashes: CallCacheHashes, response: SucceededResponse)(implicit ec: ExecutionContext): Future[Unit] = {
+    val preparation = Try {
+      val metainfo = CallCachingResultMetainfoEntry(
+        workflowUuid = workflowId.toString,
+        callFqn = response.jobKey.call.fullyQualifiedName,
+        scatterIndex = response.jobKey.index.fromIndex,
+        returnCode = response.returnCode,
+        allowResultReuse = true,
+        callCachingResultMetaInfoEntryId = None)
+
+      val hashes = callCacheHashes.hashes map { hash => HashKeyAndValue(hash.hashKey.key, hash.hashValue.value) }
+      val result = toResultSimpletons(response.jobOutputs)
+
+      (metainfo, hashes, result)
+    }
+
+    Future.fromTry(preparation) map { case (metainfo, hashes, result) =>
+      database.addToCache(metainfo, hashes, result)
```

Comment:
Is this whole thing thread safe? If there are 1000 simultaneous calls to CallCache.addToCache are there any potential concurrency issues?